src/v/pandaproxy/sharded_client_cache.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "config/broker_authn_endpoint.h"
+#include "hashing/jump_consistent_hash.h"
+#include "hashing/xx.h"
+#include "pandaproxy/kafka_client_cache.h"
+#include "pandaproxy/logger.h"
+#include "pandaproxy/types.h"
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/timer.hh>
+
+#include <chrono>
+#include <utility>
+
+namespace pandaproxy {
+
+inline ss::shard_id user_shard(const ss::sstring& name) {
+    auto hash = xxhash_64(name.data(), name.length());
+    return jump_consistent_hash(hash, ss::smp::count);
+}
+
+class sharded_client_cache {
+public:
+    ss::future<> start(
+      ss::smp_service_group sg,
+      YAML::Node const& cfg,
+      config::binding<size_t> max_size,
+      config::binding<std::chrono::milliseconds> keep_alive);
+
+    ss::future<> stop();
+
+    static std::chrono::milliseconds clean_timer();
+
+    template<
+      typename Func,
+      typename... Args,
+      typename Ret = ss::futurize_t<
+        std::invoke_result_t<Func, kafka_client_cache&, Args...>>>
+    Ret invoke_on_cache(credential_t const& user, Func&& func, Args&&... args) {
+        // Access the cache on the appropriate shard.
+        ss::shard_id u_shard{user_shard(user.name)};
+        return ss::with_gate(_gate, [this, u_shard, &func, &args...] {
```

Comment:
should the capture list be `func = std::forward<Func>(func), args = std::tuple(std::move(args)...)` ? (i looked at the seastar impl of invoke_on which forwards args like this through an intermediate lambda)

Reply:
> should the capture list be `func = std::forward<Func>(func), args = std::tuple(std::move(args)...)` ? (i looked at the seastar impl of invoke_on which forwards args like this through an intermediate lambda)

That is was the way to capture parameter packs prior to C++20, but now with C++20 you can use parameter packs in the capture list. I tried to find a place C++ docs to reference this but [here is a blog post](https://oleksandrkvl.github.io/2021/04/02/cpp-20-overview.html) that shows this (Search for heading "Pack expansion in lambda init-capture")

Reply:
Found it in C++ docs https://en.cppreference.com/w/cpp/language/parameter_pack
Search for "Lambda captures" heading

Reply:
I think the capture converts all the args to an lvalue reference.

Perhaps it should look like this:
```c++
        return ss::with_gate(
          _gate,
          [this,
           u_shard,
           func{std::forward<Func>(func)},
           ... args{std::forward<Args>(args)}]() mutable {
              return _cache.invoke_on(
                u_shard,
                _smp_opts,
                std::forward<Func>(func),
                std::forward<Args>(args)...);
```
In reality it probably makes sense to drop `Args` altogether.

Reply:
Ignore me, you're right:
```c++
#include <memory>
#include <iostream>

template<typename Func, typename... Args>
auto forward(Func&& func,Args&& ...args) {
    return [func{std::forward<Func>(func)}, &args...]() mutable {
        return std::forward<Func>(func)(std::forward<Args>(args)...);
    };
}

int rvalue(std::unique_ptr<int>&& p) { return *p; };
int lvalue(std::unique_ptr<int> const& p) { return *p; };


int main() {
    std::cout << "rvalue: " << forward(&rvalue, std::make_unique<int>(42))() << "\n";
    auto p = std::make_unique<int>(42);
    std::cout << "lvalue: " << forward(&lvalue, p)();
}
```